src/inscription.rs

```diff
@@ -0,0 +1,377 @@
+use {
+  crate::{Ordinal, Transaction, VecDeque},
+  bitcoin::{
+    blockdata::{
+      opcodes,
+      script::{self, Instruction},
+    },
+    util::taproot::TAPROOT_ANNEX_PREFIX,
+    Script, Witness,
+  },
+  std::str::{self, Utf8Error},
+};
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct Inscription(pub(crate) String);
+
+impl Inscription {
+  pub(crate) fn from_witness(witness: &Witness) -> Option<Self> {
+    InscriptionParser::parse(witness).ok()
+  }
+
+  pub(crate) fn from_transaction(
+    tx: &Transaction,
+    input_ordinal_ranges: &VecDeque<(u64, u64)>,
+  ) -> Option<(Ordinal, Inscription)> {
+    if let Some(tx_in) = tx.input.get(0) {
+      if let Some(inscription) = Inscription::from_witness(&tx_in.witness) {
+        if let Some((start, _end)) = input_ordinal_ranges.get(0) {
+          return Some((Ordinal(*start), inscription));
+        }
+      }
+    }
+
+    None
+  }
+}
+
+#[derive(Debug, PartialEq)]
+enum Error {
+  EmptyWitness,
+  KeyPathSpend,
+  Script(script::Error),
+  NoInscription,
+  Utf8Decode(Utf8Error),
+  InvalidInscription,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+struct InscriptionParser<'a> {
+  next: usize,
+  instructions: Vec<Instruction<'a>>,
+}
+
+impl<'a> InscriptionParser<'a> {
+  fn parse(witness: &Witness) -> Result<Inscription> {
+    let mut witness = witness.to_vec();
```

Comment:
We need to remove this for performance reasons. Converting every witness we encounter to a vec will be very bad for performance.